pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexDeclaration.java

```diff
@@ -0,0 +1,114 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.segment.spi.index;
+
+import com.google.common.base.Preconditions;
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+
+// This should be a sealed class in modern Java
+public class IndexDeclaration<C> {
```

Comment:
Do we need to know if an index is declared or default? In what scenario are we going to get an undeclared index?

Reply:
In this partial PR doesn't seem to be needed, but it is needed in the PR with all the changes, mainly to:
1. Be able to correctly merge new and old syntax
2. Remove undeclared values default values when config is serialized as JSON.

Also, it doesn't hurt. We could model the config as not null when enabled and null when disabled, but I think it is better to have a model with semantic values where disabled is modeled with an actual object. Once you do that, you need the `IndexDeclaration` indirection, so we need to pay for the object.